src/XR/features/WebXRNearInteraction.ts

```diff
@@ -0,0 +1,590 @@
+            };
+
+            this._xrSessionManager.session.addEventListener("selectstart", selectStartListener);
+            this._xrSessionManager.session.addEventListener("selectend", selectEndListener);
+        }
+    }
+
+    private _detachController(xrControllerUniqueId: string) {
+        const controllerData = this._controllers[xrControllerUniqueId];
+        if (!controllerData) {
+            return;
+        }
+        if (controllerData.squeezeComponent) {
+            if (controllerData.onSqueezeButtonChangedObserver) {
+                controllerData.squeezeComponent.onButtonStateChangedObservable.remove(controllerData.onSqueezeButtonChangedObserver);
+            }
+        }
+        if (controllerData.selectionComponent) {
+            if (controllerData.onButtonChangedObserver) {
+                controllerData.selectionComponent.onButtonStateChangedObservable.remove(controllerData.onButtonChangedObserver);
+            }
+        }
+        if (controllerData.onFrameObserver) {
+            this._xrSessionManager.onXRFrameObservable.remove(controllerData.onFrameObserver);
+        }
+        if (controllerData.eventListeners) {
+            Object.keys(controllerData.eventListeners).forEach((eventName: string) => {
+                const func = controllerData.eventListeners && controllerData.eventListeners[eventName as XREventType];
+                if (func) {
+                    this._xrSessionManager.session.removeEventListener(eventName as XREventType, func);
+                }
+            });
+        }
+        controllerData.pickIndexMeshTip?.dispose();
+        controllerData.hoverIndexMeshTip?.dispose();
+
+        // remove from the map
+        delete this._controllers[xrControllerUniqueId];
+        if (this._attachedController === xrControllerUniqueId) {
+            // check for other controllers
+            const keys = Object.keys(this._controllers);
+            if (keys.length) {
+                this._attachedController = keys[0];
+            } else {
+                this._attachedController = "";
+            }
+        }
+    }
+
+    private _generateNewHandTipMeshes() {
+        // populate information for near hover, pick and pinch
+        const meshCreationScene = this._options.useUtilityLayer ? this._options.customUtilityLayerScene || UtilityLayerRenderer.DefaultUtilityLayer.utilityLayerScene : this._scene;
+        var hoverIndexMeshTip = null;
+        var pickIndexMeshTip = null;
+        hoverIndexMeshTip = SphereBuilder.CreateSphere("IndexHoverSphere", { diameter: 1 }, meshCreationScene);
+        hoverIndexMeshTip.scaling.set(this._hoverRadius, this._hoverRadius, this._hoverRadius);
+        hoverIndexMeshTip.isVisible = false;
+
+        pickIndexMeshTip = SphereBuilder.CreateSphere("IndexPickSphere", { diameter: 1 }, meshCreationScene);
+        pickIndexMeshTip.scaling.set(this._pickRadius, this._pickRadius, this._pickRadius);
+        pickIndexMeshTip.isVisible = false;

```

Comment:
nit: It's small-scale, but since these two chunks are pretty much the same, they could be lambda'd